internal/bminventory/inventory.go

```diff
@@ -3657,11 +3653,28 @@ func (b *bareMetalInventory) GetPresignedForClusterFiles(ctx context.Context, pa
 }
 
 func (b *bareMetalInventory) DownloadClusterFiles(ctx context.Context, params installer.DownloadClusterFilesParams) middleware.Responder {
-	respBody, contentLength, err := b.DownloadClusterFilesInternal(ctx, params)
-	if err != nil {
-		return common.GenerateErrorResponder(err)
+	log := logutil.FromContext(ctx, b.log)
+
+	if params.FileName == "discovery.ign" {
+		infraEnv, err := common.GetInfraEnvFromDB(b.db, params.ClusterID)
+		if err != nil {
+			return common.GenerateErrorResponder(err)
+		}
+
+		cfg, err := b.IgnitionBuilder.FormatDiscoveryIgnitionFile(infraEnv, b.IgnitionConfig, false, b.authHandler.AuthType())
+		if err != nil {
+			log.WithError(err).Error("Failed to format ignition config")
+			return common.GenerateErrorResponder(err)
+		}
+
+		return installer.NewDownloadClusterFilesOK().WithPayload(ioutil.NopCloser(strings.NewReader(cfg)))
```

Comment:
I think we should also continue to use the `filemiddleware` here.

Reply:
Fixed

```
$ curl -v "$ASSISTED_SERVICE_IP:$ASSISTED_SERVICE_PORT/api/assisted-install/v1/clusters/$CLUSTER_ID/downloads/files?file_name=discovery.ign"
*   Trying 192.168.39.219:31426...
* Connected to 192.168.39.219 (192.168.39.219) port 31426 (#0)
> GET /api/assisted-install/v1/clusters/6aae6063-5250-4e75-96f3-d3e9af16f33f/downloads/files?file_name=discovery.ign HTTP/1.1
> Host: 192.168.39.219:31426
> User-Agent: curl/7.71.1
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Content-Disposition: attachment; filename="discovery.ign"
< Content-Length: 8370
< Content-Type: application/octet-stream
< Date: Wed, 28 Jul 2021 21:29:30 GMT
<
{
  "ignition": {
    "version": "3.1.0"
  },
  "passwd": {
    "users": [
      {"groups":["sudo"],"name":"core","passwordHash":"!","sshAuthorizedKeys":["..."]}
    ]
  },
  "systemd": {
    "units": [{
      "name": "agent.service",
      "enabled": true,
      "contents": "[Service]\nType=simple\nRestart=always\nRestartSec=3\nStartLimitInterval=0\nEnvironment=HTTP_PROXY=\nEnvironment=http_proxy=\nEnvironment=HTTPS_PROXY=\nEnvironment=https_proxy=\nEnvironment=NO_PROXY=\nEnvironment=no_proxy=\nTimeoutStartSec=180\nExecStartPre=/usr/local/bin/agent-fix-bz1964591 quay.io/ocpmetal/assisted-installer-agent:latest\nExecStartPre=podman run --privileged --rm -v /usr/local/bin:/hostbin quay.io/ocpmetal/assisted-installer-agent:latest cp /usr/bin/agent /hostbin\nExecStart=/usr/local/bin/agent --url http://<pending>:8090 --cluster-id 6aae6063-5250-4e75-96f3-d3e9af16f33f --agent-version quay.io/ocpmetal/assisted-installer-agent:latest --insecure=true  \n\n[Unit]\nWants=network-online.target\nAfter=network-online.target\n\n[Install]\nWantedBy=multi-user.target"
    },
    {
        "name": "selinux.service",
        "enabled": true,
        "contents": "[Service]\nType=oneshot\nExecStartPre=checkmodule -M -m -o /root/assisted.mod /root/assisted.te\nExecStartPre=semodule_package -o /root/assisted.pp -m /root/assisted.mod\nExecStart=semodule -i /root/assisted.pp\n\n[Install]\nWantedBy=multi-user.target"
    }
    ]
  },
  "storage": {
    "files": [{
      "overwrite": true,
      "path": "/usr/local/bin/agent-fix-bz1964591",
      "mode": 755,
      "user": {
          "name": "root"
      },
      "contents": { "source": "data:,%23%21%2Fusr%2Fbin%2Fsh%0A%0A%23%20This%20script%20is%20a%20workaround%20for%20bugzilla%201964591%20where%20symlinks%20inside%20%2Fvar%2Flib%2Fcontainers%2F%20get%0A%23%20corrupted%20under%20some%20circumstances.%0A%23%0A%23%20In%20order%20to%20let%20agent.service%20start%20correctly%20we%20are%20checking%20here%20whether%20the%20requested%0A%23%20container%20image%20exists%20and%20in%20case%20%22podman%20images%22%20returns%20an%20error%20we%20try%20removing%20the%20faulty%0A%23%20image.%0A%23%0A%23%20In%20such%20a%20scenario%20agent.service%20will%20detect%20the%20image%20is%20not%20present%20and%20pull%20it%20again.%20In%20case%0A%23%20the%20image%20is%20present%20and%20can%20be%20detected%20correctly%2C%20no%20any%20action%20is%20required.%0A%0AIMAGE=$%28echo%20$1%20%7C%20sed%20%27s%2F:.%2A%2F%2F%27%29%0Apodman%20images%20%7C%20grep%20$IMAGE%20%7C%7C%20podman%20rmi%20--force%20$1%20%7C%7C%20true%0A" }
    },
    {
      "overwrite": true,
      "path": "/etc/motd",
      "mode": 420,
      "user": {
          "name": "root"
      },
      "contents": { "source": "data:,%0A%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%0AThis%20is%20a%20host%20being%20installed%20by%20the%20OpenShift%20Assisted%20Installer.%0AIt%20will%20be%20installed%20from%20scratch%20during%20the%20installation.%0A%0AThe%20primary%20service%20is%20agent.service.%20To%20watch%20its%20status%2C%20run:%0Asudo%20journalctl%20-u%20agent.service%0A%0ATo%20view%20the%20agent%20log%2C%20run:%0Asudo%20journalctl%20TAG=agent%0A%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%20%20%2A%2A%0A" }
    },
    {
      "overwrite": true,
      "path": "/etc/NetworkManager/conf.d/01-ipv6.conf",
      "mode": 420,
      "user": {
          "name": "root"
      },
      "contents": { "source": "data:,%0A%5Bconnection%5D%0Aipv6.dhcp-iaid=mac%0Aipv6.dhcp-duid=ll%0A" }
    },
    {
        "overwrite": true,
        "path": "/root/.docker/config.json",
        "mode": 420,
        "user": {
            "name": "root"
        },
        "contents": { "source": "..." }
    },
    {
        "overwrite": true,
        "path": "/root/assisted.te",
        "mode": 420,
        "user": {
            "name": "root"
        },
        "contents": { "source": "data:text/plain;base64,Cm1vZHVsZSBhc3Npc3RlZCAxLjA7CnJlcXVpcmUgewogICAgICAgIHR5cGUgY2hyb255ZF90OwogICAgICAgIHR5cGUgY29udGFpbmVyX2ZpbGVfdDsKICAgICAgICB0eXBlIHNwY190OwogICAgICAgIGNsYXNzIHVuaXhfZGdyYW1fc29ja2V0IHNlbmR0bzsKICAgICAgICBjbGFzcyBkaXIgc2VhcmNoOwogICAgICAgIGNsYXNzIHNvY2tfZmlsZSB3cml0ZTsKfQojPT09PT09PT09PT09PSBjaHJvbnlkX3QgPT09PT09PT09PT09PT0KYWxsb3cgY2hyb255ZF90IGNvbnRhaW5lcl9maWxlX3Q6ZGlyIHNlYXJjaDsKYWxsb3cgY2hyb255ZF90IGNvbnRhaW5lcl9maWxlX3Q6c29ja19maWxlIHdyaXRlOwphbGxvdyBjaHJvbnlkX3Qgc3BjX3Q6dW5peF9kZ3JhbV9zb2NrZXQgc2VuZHRvOwo=" }
    }]
  }
* Connection #0 to host 192.168.39.219 left intact
}%
```